cpp/dolfin/fem/DofMapPermuter.cpp

```diff
@@ -76,47 +76,42 @@ std::array<int, 4> _calculate_tetrahedron_orders(int v1, int v2, int v3, int v4)
 //-----------------------------------------------------------------------------
 /// Makes a permutation to flip the dofs on an edge
 /// @param[in] edge_dofs Number of edge dofs
+/// @param[in] blocksize Block size
 /// @return The permutation to reverse the dofs
-std::vector<int> _edge_flip(const int edge_dofs)
+std::vector<int> _edge_flip(const int edge_dofs, const int blocksize)
 {
   // This will only be called at most once for each mesh
+  const int blocks = edge_dofs / blocksize;
   int j = 0;
   std::vector<int> flip(edge_dofs);
-  for (int i = 0; i < edge_dofs; ++i)
-    flip[j++] = edge_dofs - (i + 1);
+  for (int i = 0; i < blocks; ++i)
+    for (int k = 0; k < blocksize; ++k)
+      flip[j++] = edge_dofs - (i + 1) * blocksize + k;
   return flip;
 }
 //-----------------------------------------------------------------------------
 /// Makes permutations to rotate and reflect the dofs on a triangle
 /// @param[in] edge_dofs Number of dofs on the face of the triangle
 /// @return Permutations to rotate and reflect the dofs
 std::array<std::vector<int>, 2>
-_triangle_rotation_and_reflection(const int face_dofs)
+_triangle_rotation_and_reflection(const int face_dofs, const int blocksize)
 {
   // This will only be called at most once for each mesh
-  float root = std::sqrt(8 * face_dofs + 1);
-
-  // FIXME: If the dofs are not arranged in a triangle (eg N1curl spaces), what
-  // should this do?
-  if (root != std::floor(root) || int(root) % 2 != 1)
-  {
-    std::vector<int> trivial(face_dofs);
-    std::iota(trivial.begin(), trivial.end(), 0);
-    return {trivial, trivial};
-  }
-
+  const int blocks = face_dofs / blocksize;
+  float root = std::sqrt(8 * blocks + 1);
   assert(root == std::floor(root) && int(root) % 2 == 1);
   int side_length = (root - 1) / 2; // side length of the triangle of face dofs
 
   std::vector<int> rotation(face_dofs);
   {
     int j = 0;
     int i = 1;
-    for (int st = face_dofs - 1; st >= 0; st -= (i++))
+    for (int st = blocks - 1; st >= 0; st -= (i++))
```

Comment:
I find this kind of update confusing. Can we just put `++i;` somewhere in the loop instead?